recipes/LibriSpeech/SSL/train.py

```diff
@@ -0,0 +1,398 @@
+        should_step = self.step % self.grad_accumulation_factor == 0
+        # Managing automatic mixed precision
+        if self.auto_mix_prec:
+            with self.modules.model.no_sync() if not should_step else ExitStack():
+                with torch.cuda.amp.autocast():
+                    outputs = self.compute_forward(batch, Stage.TRAIN)
+                    objectives = self.compute_objectives(
+                        outputs, batch, Stage.TRAIN
+                    )
+                loss = objectives["loss"]
+                if self.hparams.diversity_loss_weight == 0.0:
+                    backprop_loss = loss
+                else:
+                    backprop_loss = (
+                        loss
+                        + objectives["diversity_loss"] * self.hparams.diversity_loss_weight# * objectives["num_masked"]
+                        + objectives["latent_l2"] * self.hparams.latent_l2_penalty_weight
+                    )
+                self.scaler.scale(
+                    backprop_loss / self.grad_accumulation_factor
+                ).backward()
+
+                objectives["total_loss"] = backprop_loss.detach()
+                if should_step:
+                    self.scaler.unscale_(self.optimizer)
+                    if self.check_gradients(loss):
+                        self.scaler.step(self.optimizer)
+                    self.scaler.update()
+                    self.optimizer_step += 1
+                    self.optimizer.zero_grad()
+        else:
+
+            with self.modules.model.no_sync() if not should_step else ExitStack():
+                outputs = self.compute_forward(batch, Stage.TRAIN)
+                objectives = self.compute_objectives(
+                    outputs, batch, Stage.TRAIN
+                )
+                loss = objectives["loss"]
+                if self.hparams.diversity_loss_weight == 0.0:
+                    backprop_loss = loss
+                else:
+                    backprop_loss = (
+                        loss
+                        + objectives["diversity_loss"] * self.hparams.diversity_loss_weight
+                        + objectives["latent_l2"] * self.hparams.latent_l2_penalty_weight
+                    )
+                (backprop_loss / self.grad_accumulation_factor).backward()
+                objectives["total_loss"] = backprop_loss.detach()
+                if should_step:
+                    self.optimizer.step()
+                    self.optimizer.zero_grad()
+                    self.optimizer_step += 1
+
+        if should_step:
+            if isinstance(
+                self.modules.model, DistributedDataParallel
+            ):
+                w2v_model = self.modules.model.module
+            else:
+                w2v_model = self.modules.model
+            if w2v_model.quantiser is not None:
+                w2v_model.quantiser.update_temp(
+                    self.optimizer_step
+                )
+
+            if (
```

Comment:
Ok I understand the need for logging here. What I propose is to introduce something bigger directly in the core.py of SpeechBrain. I think that logging at every batch totally is relevant for many recipes. What about creating a callback in the official Brain class that is called at the end of fit_batch ? Then you could override it here to manage the logging.

Reply:
Yes I like that idea (I agree putting logging in fit_batch is awkward), will do